### PR TITLE
refactor(DeclareContract): revamp RPC starkNet_declareContract

### PR DESCRIPTION
**Pull request by khanti42**
This PR is to revamp the RPC starkNet_declareContract handler to align with new RPC standard

It also fixes the `function_idx` type in `SierraContractEntryPointFieldsStruct` of superstruct validation from `string` to `number`.

BEGIN_COMMIT_OVERRIDE
feat: Refactor RPC `starkNet_declareContract` to have superstruct validation
END_COMMIT_OVERRIDE

### CODE REVIEW

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  protected responseStruct = DeclareContractResponseStruct;
+
+  protected async preExecute(params: DeclareContractParams): Promise<void> {
+    // Define mappings to ensure backward compatibility with previous versions of the API.
+    // These mappings replace deprecated parameter names with the updated equivalents,
+    // allowing older integrations to function without changes
+    const paramMappings: Record<string, string> = {
+      senderAddress: 'address',
+      invocationsDetails: 'details',
+    };
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:59]: nit:

if we accept above change, that also required to update too
```suggestion
    const paramMappings: Record<string, string> = {
      senderAddress: 'address',
      invocationsDetails: 'details',
      contractPayload: 'payload'
    };
```

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    mapDeprecatedParams(params, paramMappings);
+    await super.preExecute(params);
+  }
+
+  /**
+   * Execute the declare contract request handler.
+   * It will show a confirmation dialog to the user before signing the contract declaration.
+   *
+   * @param params - The parameters of the request.
+   * @returns the signature of the message in string array.
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:71]:
```suggestion
   * @param params - The parameters of the request.
   * @param params.address - The address of the request account.
   * @param params.payload - The contract payload of the declare transaction.
   * @param [params.details] - The declare transaction details.
   * @param params.chainId - The chain id of the network.
   * @returns A Promise that resolve the `DeclareContractResponse` object.

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    components.push(
+      row(
+        'Signer Address',
+        text({
+          value: address,
+          markdown: false,
+        }),
+      ),
+    );
+
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:111]: lets add a divider in between all components
```
components.push(divider());
```

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+          text({
+            value: casmDetails,
+            markdown: false,
+          }),
+        ),
+      );
+    }
+
+    // Add Max Fee in ETH, Nonce, and Version if they are available in details
+    if (details?.maxFee) {
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:179]: i just realize we didnt estimate the fee when delcare a transaction in previous code
im not sure shall we add it back, else the user will always need to pay by ETH
(we can do it in other PR)

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      components.push(
+        row(
+          'Version',
+          text({
+            value: details.version.toString(),
+            markdown: false,
+          }),
+        ),
+      );
+    }
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:214]:
```suggestion
```

May we we should align with execute-txn?

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        }),
+      ),
+    );
+
+    // Add contract details, compiled class hash, and class hash
+    if (contractPayload.contract) {
+      const contractDetails =
+        typeof contractPayload.contract === 'string'
+          ? contractPayload.contract
+          : JSON.stringify(contractPayload.contract, null, 2);
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:128]:
```suggestion
          : toJson(contractPayload.contract);
```

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            value: contractPayload.classHash,
+            markdown: false,
+          }),
+        ),
+      );
+    }
+
+    // Add Casm details if available
+    if (contractPayload.casm) {
+      const casmDetails = JSON.stringify(contractPayload.casm, null, 2);
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:166]:
```suggestion
      const casmDetails = toJson(contractPayload.casm);
```

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      this.account.privateKey,
+      contractPayload,
+      details,
+    )) as DeclareContractResponse;
+  }
+
+  protected async getDeclareContractConsensus(params: DeclareContractParams) {
+    const { contractPayload, details, address } = params;
+    const components: Component[] = [];
+
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:100]: missing header 
```suggestion
   components.push(heading('Do you want to sign this transaction?'))
```

File: packages/starknet-snap/src/rpcs/declare-contract.ts [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    // Add the signer address
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.ts:101]: nit:

we can remove those comments
```suggestion
```

**Review by stanleyyconsensys — COMMENTED** [review]
Mostly look good to me, just may have some change on the dialog render

File: packages/starknet-snap/src/rpcs/declare-contract.ts (hunk 4 above)
Comment (khanti42) [reply packages/starknet-snap/src/rpcs/declare-contract.ts:187]: I think we could add it here. But then I am thinking we should maybe consider that any action that requires to pay fees (e.g. execute, and declare) should also deploy the account if not. Wdyt ?

File: packages/starknet-snap/src/rpcs/declare-contract.test.ts [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    );
+    confirmDialogSpy.mockResolvedValue(false);
+
+    await expect(declareContract.execute(request)).rejects.toThrow(
+      UserRejectedOpError,
+    );
+  });
+
+  it('throws `InvalidRequestParamsError` when request parameter is not correct', async () => {
+    await expect(declareContract.execute({} as unknown as any)).rejects.toThrow(
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.test.ts:139]:
```suggestion
    await expect(declareContract.execute({} as unknown as DeclareContractParams)).rejects.toThrow(
```

File: packages/starknet-snap/src/rpcs/declare-contract.test.ts [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+        },
+      },
+      {
+        type: 'divider',
+      },
+      {
+        type: 'row',
+        label: 'Network',
+        value: {
+          value: STARKNET_SEPOLIA_TESTNET_NETWORK.name,
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.test.ts:217]: we can get the network from prepareMockDeclareContract

File: packages/starknet-snap/src/rpcs/declare-contract.test.ts [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+    await declareContract.execute(request);
+
+    const calls = confirmDialogSpy.mock.calls[0][0];
+    expect(calls).toStrictEqual([
+      {
+        type: 'heading',
+        value: 'Do you want to sign this transaction?',
+      },
+      {
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.test.ts:201]: we can use buildRowComponent, buildDividerComponent from test helper

File: packages/starknet-snap/src/rpcs/declare-contract.test.ts [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    const maxFeeInEth = utils.formatUnits(details.maxFee, 'ether');
+    const transactionHash = '0x123';
+
+    const { request, confirmDialogSpy, account } =
+      await prepareMockDeclareContract(transactionHash, payload, details);
+
+    await declareContract.execute(request);
+
+    const calls = confirmDialogSpy.mock.calls[0][0];
+    expect(calls).toStrictEqual([
```
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/rpcs/declare-contract.test.ts:196]:
```suggestion
    expect(confirmDialogSpy).toHaveBeenCalledWith([
```

**Review by stanleyyconsensys — COMMENTED** [review]
LGTM, the change mostly on test

**Review by stanleyyconsensys — APPROVED** [review]
LGTM

**Review by stanleyyconsensys — APPROVED** [review]
LGTM

**Pull request merged**